File: src/main/java/com/nurkiewicz/tsclass/parser/visitors/MemberExpressionVisitor.java
package com.nurkiewicz.tsclass.parser.visitors;

import com.nurkiewicz.tsclass.antlr.parser.TypeScriptBaseVisitor;
import com.nurkiewicz.tsclass.antlr.parser.TypeScriptParser;
import com.nurkiewicz.tsclass.parser.ast.expr.Expression;
import com.nurkiewicz.tsclass.parser.ast.expr.Identifier;
import com.nurkiewicz.tsclass.parser.ast.expr.NumberLiteral;

class MemberExpressionVisitor extends TypeScriptBaseVisitor<Expression> {
        @Override
        public Expression visitPrimaryExpression(TypeScriptParser.PrimaryExpressionContext ctx) {
            if (ctx.openParen() != null) {
                return ctx.expression().accept(new ExpressionVisitor());
            }
            if (ctx.IDENT() != null) {
                return new Identifier(ctx.IDENT().getText());
            }
            return ctx.accept(new TypeScriptBaseVisitor<Expression>() {
                @Override
                public Expression visitLiteral(TypeScriptParser.LiteralContext ctx) {
                    return new NumberLiteral(Double.valueOf(ctx.NUMERIC_LITERAL().getText()));
                }
            });
        }
    }

File: src/main/kotlin/com/nurkiewicz/tsclass/parser/visitors/MemberExpressionVisitor.kt
package com.nurkiewicz.tsclass.parser.visitors

import com.nurkiewicz.tsclass.antlr.parser.TypeScriptBaseVisitor
import com.nurkiewicz.tsclass.antlr.parser.TypeScriptParser
import com.nurkiewicz.tsclass.parser.ast.Expression
import com.nurkiewicz.tsclass.parser.ast.Identifier
import com.nurkiewicz.tsclass.parser.ast.NumberLiteral

internal class MemberExpressionVisitor : TypeScriptBaseVisitor<Expression>() {
    override fun visitPrimaryExpression(ctx: TypeScriptParser.PrimaryExpressionContext): Expression {
        if (ctx.openParen() != null) {
            return ctx.expression().accept(ExpressionVisitor)
        }
        return if (ctx.IDENT() != null) {
            Identifier(ctx.IDENT().text)
        } else ctx.accept(object : TypeScriptBaseVisitor<Expression>() {
            override fun visitLiteral(ctx: TypeScriptParser.LiteralContext): Expression {
                return NumberLiteral(ctx.NUMERIC_LITERAL().text.toDouble())
            }
        })
    }
}